File: scripts/ai/tetrisAI.js
// Tetris AI - Intelligent opponent for Battle Mode
import { GRID_WIDTH, GRID_HEIGHT } from '../grid.js';

export class TetrisAI {
    constructor(difficulty = 3) {
        this.difficulty = difficulty; // 1-5 (Easy to Grandmaster)
        this.grid = this.createEmptyGrid();
        this.currentPiece = null;
        this.nextPieces = [];
        this.score = 0;
        this.lines = 0;
        this.level = 1;
        this.garbageQueue = [];
        
        // AI parameters based on difficulty
        this.setupDifficultyParams();
        
        // Decision making
        this.thinkingTime = 0;
        this.currentMove = null;
        this.moveQueue = [];
    }
    
    setupDifficultyParams() {
        switch (this.difficulty) {
            case 1: // Easy
                this.params = {
                    thinkingDelay: 800,
                    mistakeRate: 0.3,
                    lookAhead: 0,
                    evaluationWeights: {
                        height: -0.5,
                        holes: -3,
                        bumpiness: -0.2,
                        lines: 10,
                        wells: 0
                    },
                    dropSpeed: 1000,
                    garbageDefenseSkill: 0.3
                };
                break;
                
            case 2: // Normal
                this.params = {
                    thinkingDelay: 500,
                    mistakeRate: 0.15,
                    lookAhead: 1,
                    evaluationWeights: {
                        height: -0.8,
                        holes: -5,
                        bumpiness: -0.5,
                        lines: 15,
                        wells: 0.5
                    },
                    dropSpeed: 700,
                    garbageDefenseSkill: 0.5
                };
                break;
                
            case 3: // Hard
                this.params = {
                    thinkingDelay: 300,
                    mistakeRate: 0.08,
                    lookAhead: 1,
                    evaluationWeights: {
                        height: -1,
                        holes: -8,
                        bumpiness: -0.8,
                        lines: 20,
                        wells: 1,
                        tspinSetup: 5
                    },
                    dropSpeed: 400,
                    garbageDefenseSkill: 0.7
                };
                break;
                
            case 4: // Expert
                this.params = {
                    thinkingDelay: 150,
                    mistakeRate: 0.03,
                    lookAhead: 2,
                    evaluationWeights: {
                        height: -1.2,
                        holes: -10,
                        bumpiness: -1,
                        lines: 25,
                        wells: 2,
                        tspinSetup: 8,
                        tetrisWell: 3
                    },
                    dropSpeed: 200,
                    garbageDefenseSkill: 0.85
                };
                break;
                
            case 5: // Grandmaster
                this.params = {
                    thinkingDelay: 50,
                    mistakeRate: 0.01,
                    lookAhead: 3,
                    evaluationWeights: {
                        height: -1.5,
                        holes: -15,
                        bumpiness: -1.2,
                        lines: 30,
                        wells: 3,
                        tspinSetup: 10,
                        tetrisWell: 5,
                        perfectClear: 50
                    },
                    dropSpeed: 100,
                    garbageDefenseSkill: 0.95
                };
                break;
        }
    }
    
    createEmptyGrid() {
        return Array(GRID_HEIGHT).fill(null).map(() => Array(GRID_WIDTH).fill(0));
    }
    
    update(deltaTime) {
        this.thinkingTime += deltaTime;
        
        // Process garbage lines
        if (this.garbageQueue.length > 0 && Math.random() < this.params.garbageDefenseSkill) {
            this.processGarbage();
        }
        
        // Make decision if enough thinking time has passed
        if (this.thinkingTime >= this.params.thinkingDelay) {
            this.thinkingTime = 0;
            
            if (!this.currentMove && this.currentPiece) {
                this.decideBestMove();
            }
            
            if (this.currentMove) {
                this.executeNextMove();
            }
        }
    }
    
    setPiece(piece) {
        this.currentPiece = piece;
        this.currentMove = null;
        this.moveQueue = [];
    }
    
    decideBestMove() {
        if (!this.currentPiece) return;
        
        const moves = this.getAllPossibleMoves(this.currentPiece);
        let bestMove = null;
        let bestScore = -Infinity;
        
        for (let move of moves) {
            const score = this.evaluateMove(move);
            
            // Add randomness based on difficulty
            const randomFactor = (Math.random() - 0.5) * (5 - this.difficulty) * 10;
            const finalScore = score + randomFactor;
            
            if (finalScore > bestScore) {
                bestScore = finalScore;
                bestMove = move;
            }
        }
        
        // Chance to make a mistake
        if (Math.random() < this.params.mistakeRate) {
            bestMove = moves[Math.floor(Math.random() * moves.length)];
        }
        
        this.currentMove = bestMove;
        if (bestMove) {
            this.generateMoveSequence(bestMove);
        }
    }
    
    getAllPossibleMoves(piece) {
        const moves = [];
        const rotations = 4; // All possible rotations
        
        for (let rotation = 0; rotation < rotations; rotation++) {
            const rotatedPiece = this.rotatePiece(piece, rotation);
            
            for (let x = -2; x < GRID_WIDTH + 2; x++) {
                const move = {
                    x: x,
                    rotation: rotation,
                    piece: rotatedPiece
                };
                
                if (this.isValidPosition(rotatedPiece, x)) {
                    // Find drop position
                    move.y = this.getDropPosition(rotatedPiece, x);
                    moves.push(move);
                }
            }
        }
        
        return moves;
    }
    
    rotatePiece(piece, times) {
        let rotated = { ...piece };
        for (let i = 0; i < times; i++) {
            rotated = this.rotateMatrix(rotated);
        }
        return rotated;
    }
    
    rotateMatrix(piece) {
        // Simple rotation logic
        const matrix = piece.shape;
        const n = matrix.length;
        const rotated = Array(n).fill(null).map(() => Array(n).fill(0));
        
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                rotated[j][n - 1 - i] = matrix[i][j];
            }
        }
        
        return { ...piece, shape: rotated };
    }
    
    isValidPosition(piece, x, y = 0) {
        // Check if piece fits at position
        for (let row = 0; row < piece.shape.length; row++) {
            for (let col = 0; col < piece.shape[row].length; col++) {
                if (piece.shape[row][col]) {
                    const gridX = x + col;
                    const gridY = y + row;
                    
                    if (gridX < 0 || gridX >= GRID_WIDTH || 
                        gridY >= GRID_HEIGHT ||
                        (gridY >= 0 && this.grid[gridY][gridX])) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
    
    getDropPosition(piece, x) {
        let y = 0;
        while (this.isValidPosition(piece, x, y + 1)) {
            y++;
        }
        return y;
    }
    
    evaluateMove(move) {
        // Simulate placing the piece
        const testGrid = this.cloneGrid();
        this.placePieceOnGrid(testGrid, move.piece, move.x, move.y);
        
        // Clear lines
        const linesCleared = this.clearLines(testGrid);
        
        // Calculate evaluation metrics
        const height = this.calculateHeight(testGrid);
        const holes = this.calculateHoles(testGrid);
        const bumpiness = this.calculateBumpiness(testGrid);
        const wells = this.calculateWells(testGrid);
        
        // Calculate score
        let score = 0;
        score += this.params.evaluationWeights.height * height;
        score += this.params.evaluationWeights.holes * holes;
        score += this.params.evaluationWeights.bumpiness * bumpiness;
        score += this.params.evaluationWeights.lines * linesCleared;
        score += this.params.evaluationWeights.wells * wells;
        
        // Advanced scoring for higher difficulties
        if (this.difficulty >= 3) {
            if (this.params.evaluationWeights.tspinSetup) {
                score += this.params.evaluationWeights.tspinSetup * 
                         this.detectTSpinSetup(testGrid, move);
            }
            
            if (this.params.evaluationWeights.tetrisWell) {
                score += this.params.evaluationWeights.tetrisWell * 
                         this.detectTetrisWell(testGrid);
            }
        }
        
        if (this.difficulty >= 5 && this.params.evaluationWeights.perfectClear) {
            if (this.isPerfectClear(testGrid)) {
                score += this.params.evaluationWeights.perfectClear;
            }
        }
        
        // Look ahead evaluation
        if (this.params.lookAhead > 0 && this.nextPieces.length > 0) {
            score += this.lookAheadEvaluation(testGrid, 0) * 0.5;
        }
        
        return score;
    }
    
    cloneGrid() {
        return this.grid.map(row => [...row]);
    }
    
    placePieceOnGrid(grid, piece, x, y) {
        for (let row = 0; row < piece.shape.length; row++) {
            for (let col = 0; col < piece.shape[row].length; col++) {
                if (piece.shape[row][col]) {
                    const gridY = y + row;
                    const gridX = x + col;
                    if (gridY >= 0 && gridY < GRID_HEIGHT && 
                        gridX >= 0 && gridX < GRID_WIDTH) {
                        grid[gridY][gridX] = piece.shape[row][col];
                    }
                }
            }
        }
    }
    
    clearLines(grid) {
        let linesCleared = 0;
        for (let row = GRID_HEIGHT - 1; row >= 0; row--) {
            if (grid[row].every(cell => cell !== 0)) {
                grid.splice(row, 1);
                grid.unshift(Array(GRID_WIDTH).fill(0));
                linesCleared++;
                row++; // Check same row again
            }
        }
        return linesCleared;
    }
    
    calculateHeight(grid) {
        for (let row = 0; row < GRID_HEIGHT; row++) {
            if (grid[row].some(cell => cell !== 0)) {
                return GRID_HEIGHT - row;
            }
        }
        return 0;
    }
    
    calculateHoles(grid) {
        let holes = 0;
        for (let col = 0; col < GRID_WIDTH; col++) {
            let blockFound = false;
            for (let row = 0; row < GRID_HEIGHT; row++) {
                if (grid[row][col] !== 0) {
                    blockFound = true;
                } else if (blockFound) {
                    holes++;
                }
            }
        }
        return holes;
    }
    
    calculateBumpiness(grid) {
        const heights = [];
        for (let col = 0; col < GRID_WIDTH; col++) {
            let height = 0;
            for (let row = 0; row < GRID_HEIGHT; row++) {
                if (grid[row][col] !== 0) {
                    height = GRID_HEIGHT - row;
                    break;
                }
            }
            heights.push(height);
        }
        
        let bumpiness = 0;
        for (let i = 0; i < heights.length - 1; i++) {
            bumpiness += Math.abs(heights[i] - heights[i + 1]);
        }
        return bumpiness;
    }
    
    calculateWells(grid) {
        let wells = 0;
        for (let col = 0; col < GRID_WIDTH; col++) {
            for (let row = 0; row < GRID_HEIGHT; row++) {
                if (grid[row][col] === 0) {
                    const leftFilled = col === 0 || grid[row][col - 1] !== 0;
                    const rightFilled = col === GRID_WIDTH - 1 || grid[row][col + 1] !== 0;
                    if (leftFilled && rightFilled) {
                        wells++;
                    }
                }
            }
        }
        return wells;
    }
    
    detectTSpinSetup(grid, move) {
        // Simplified T-spin detection
        if (move.piece.type === 'T') {
            // Check for T-spin slot
            return 0; // Simplified for now
        }
        return 0;
    }
    
    detectTetrisWell(grid) {
        // Check if rightmost column is empty (common Tetris setup)
        let rightColumnEmpty = true;
        for (let row = 0; row < GRID_HEIGHT; row++) {
            if (grid[row][GRID_WIDTH - 1] !== 0) {
                rightColumnEmpty = false;
                break;
            }
        }
        return rightColumnEmpty ? 1 : 0;
    }
    
    isPerfectClear(grid) {
        return grid.every(row => row.every(cell => cell === 0));
    }
    
    lookAheadEvaluation(grid, depth) {
        if (depth >= this.params.lookAhead || depth >= this.nextPieces.length) {
            return 0;
        }
        // Simplified lookahead
        return 0;
    }
    
    generateMoveSequence(move) {
        this.moveQueue = [];
        
        // Add rotation moves
        for (let i = 0; i < move.rotation; i++) {
            this.moveQueue.push('rotate');
        }
        
        // Add horizontal moves
        const currentX = Math.floor(GRID_WIDTH / 2) - 1; // Starting position
        const targetX = move.x;
        const direction = targetX > currentX ? 'right' : 'left';
        const distance = Math.abs(targetX - currentX);
        
        for (let i = 0; i < distance; i++) {
            this.moveQueue.push(direction);
        }
        
        // Add drop
        this.moveQueue.push('drop');
    }
    
    executeNextMove() {
        if (this.moveQueue.length === 0) {
            this.currentMove = null;
            return null;
        }
        
        return this.moveQueue.shift();
    }
    
    addGarbageLines(count) {
        this.garbageQueue.push(count);
    }
    
    processGarbage() {
        if (this.garbageQueue.length === 0) return;
        
        const lines = this.garbageQueue.shift();
        
        // Add garbage lines to grid
        for (let i = 0; i < lines; i++) {
            this.grid.shift(); // Remove top row
            const garbageLine = Array(GRID_WIDTH).fill(1);
            // Add one hole
            const hole = Math.floor(Math.random() * GRID_WIDTH);
            garbageLine[hole] = 0;
            this.grid.push(garbageLine);
        }
    }
    
    getStats() {
        return {
            score: this.score,
            lines: this.lines,
            level: this.level,
            difficulty: this.difficulty
        };
    }
    
    reset() {
        this.grid = this.createEmptyGrid();
        this.score = 0;
        this.lines = 0;
        this.level = 1;
        this.currentPiece = null;
        this.currentMove = null;
        this.moveQueue = [];
        this.garbageQueue = [];
        this.thinkingTime = 0;
    }
}

File: android-app/app/src/main/java/com/tetris/modern/rl/ai/TetrisAI.kt
package com.tetris.modern.rl.ai

import com.tetris.modern.rl.game.Grid
import com.tetris.modern.rl.game.TetrisEngine
import com.tetris.modern.rl.game.models.Piece
import com.tetris.modern.rl.game.models.PieceType
import kotlinx.coroutines.*
import kotlin.math.*
import kotlin.random.Random

/**
 * Advanced Tetris AI with 5 difficulty levels
 * Implements T-spin detection, combo tracking, and optimal pathfinding
 */
class TetrisAI(private val difficultyLevel: Int = 2) {
    
    companion object {
        // AI Difficulty Levels
        const val LEVEL_EASY = 1      // Beginner - Makes frequent mistakes
        const val LEVEL_NORMAL = 2     // Normal - Good placement, occasional mistakes
        const val LEVEL_HARD = 3       // Hard - Very good placement, rare mistakes
        const val LEVEL_EXPERT = 4     // Expert - Near-optimal play
        const val LEVEL_MASTER = 5     // Master - Perfect play with advanced strategies
        
        // Evaluation weights for different metrics
        private val WEIGHTS = mapOf(
            "height" to -0.510066,
            "holes" to -0.35663,
            "bumpiness" to -0.184483,
            "lines" to 0.760666,
            "wells" to -0.35663,
            "deepHoles" to -0.7,
            "transitions" to -0.3,
            "tSpinSetup" to 0.4,
            "tetrisWell" to 0.3
        )
    }
    
    private var grid = Grid(TetrisEngine.GRID_WIDTH, TetrisEngine.TOTAL_HEIGHT)
    private var currentPiece: Piece? = null
    private var nextPieces = listOf<PieceType>()
    private var holdPiece: PieceType? = null
    private var canHold = true
    
    private var linesCleared = 0
    private var score = 0
    private var isGameOver = false
    
    private val thinkingDelay: Long
    private val errorRate: Float
    private val lookAheadDepth: Int
    private val useTSpins: Boolean
    private val useHold: Boolean
    private val comboTracking: Boolean
    
    init {
        // Configure AI based on difficulty level
        when (difficultyLevel) {
            LEVEL_EASY -> {
                thinkingDelay = 500L
                errorRate = 0.3f
                lookAheadDepth = 0
                useTSpins = false
                useHold = false
                comboTracking = false
            }
            LEVEL_NORMAL -> {
                thinkingDelay = 300L
                errorRate = 0.15f
                lookAheadDepth = 1
                useTSpins = false
                useHold = true
                comboTracking = false
            }
            LEVEL_HARD -> {
                thinkingDelay = 150L
                errorRate = 0.05f
                lookAheadDepth = 2
                useTSpins = true
                useHold = true
                comboTracking = true
            }
            LEVEL_EXPERT -> {
                thinkingDelay = 50L
                errorRate = 0.01f
                lookAheadDepth = 3
                useTSpins = true
                useHold = true
                comboTracking = true
            }
            LEVEL_MASTER -> {
                thinkingDelay = 10L
                errorRate = 0f
                lookAheadDepth = 4
                useTSpins = true
                useHold = true
                comboTracking = true
            }
            else -> {
                thinkingDelay = 300L
                errorRate = 0.15f
                lookAheadDepth = 1
                useTSpins = false
                useHold = true
                comboTracking = false
            }
        }
    }
    
    data class Move(
        val x: Int,
        val rotation: Int,
        val useHold: Boolean = false,
        val score: Double = 0.0,
        val isTSpin: Boolean = false
    )
    
    suspend fun update(deltaTime: Long) {
        if (isGameOver) return
        
        currentPiece?.let { piece ->
            // Add thinking delay based on difficulty
            delay(thinkingDelay)
            
            // Calculate best move
            val bestMove = findBestMove(piece)
            
            // Apply error rate
            val finalMove = if (Random.nextFloat() < errorRate) {
                // Make a mistake - choose a random move instead
                findRandomMove(piece)
            } else {
                bestMove
            }
            
            // Execute the move
            executeMove(finalMove)
        }
    }
    
    private fun findBestMove(piece: Piece): Move {
        val moves = mutableListOf<Move>()
        
        // Try all possible positions and rotations
        for (rotation in 0..3) {
            for (x in -2..TetrisEngine.GRID_WIDTH + 2) {
                val testPiece = piece.copy(x = x, rotation = rotation)
                
                if (isValidPlacement(testPiece)) {
                    // Drop the piece to find landing position
                    val landedPiece = dropPiece(testPiece)
                    
                    // Evaluate the position
                    val score = evaluatePosition(landedPiece)
                    
                    // Check for T-spin
                    val isTSpin = useTSpins && checkTSpin(landedPiece)
                    
                    moves.add(Move(x, rotation, false, score, isTSpin))
                }
            }
        }
        
        // Try using hold if available
        if (useHold && canHold && holdPiece != null) {
            val heldPieceMoves = findMovesForPiece(
                Piece(holdPiece!!, TetrisEngine.GRID_WIDTH / 2, 0, 0)
            )
            heldPieceMoves.forEach { move ->
                moves.add(move.copy(useHold = true))
            }
        }
        
        // Look ahead if configured
        if (lookAheadDepth > 0 && nextPieces.isNotEmpty()) {
            moves.forEach { move ->
                val futureScore = evaluateFuture(move, lookAheadDepth)
                moves[moves.indexOf(move)] = move.copy(score = move.score + futureScore * 0.3)
            }
        }
        
        // Prioritize T-spins and special moves
        if (useTSpins) {
            moves.filter { it.isTSpin }.forEach { move ->
                moves[moves.indexOf(move)] = move.copy(score = move.score * 1.5)
            }
        }
        
        // Return the best move
        return moves.maxByOrNull { it.score } ?: Move(4, 0)
    }
    
    private fun findRandomMove(piece: Piece): Move {
        val validMoves = mutableListOf<Move>()
        
        for (rotation in 0..3) {
            for (x in 0 until TetrisEngine.GRID_WIDTH) {
                val testPiece = piece.copy(x = x, rotation = rotation)
                if (isValidPlacement(testPiece)) {
                    validMoves.add(Move(x, rotation))
                }
            }
        }
        
        return validMoves.randomOrNull() ?: Move(4, 0)
    }
    
    private fun findMovesForPiece(piece: Piece): List<Move> {
        val moves = mutableListOf<Move>()
        
        for (rotation in 0..3) {
            for (x in 0 until TetrisEngine.GRID_WIDTH) {
                val testPiece = piece.copy(x = x, rotation = rotation)
                if (isValidPlacement(testPiece)) {
                    val score = evaluatePosition(dropPiece(testPiece))
                    moves.add(Move(x, rotation, false, score))
                }
            }
        }
        
        return moves
    }
    
    private fun evaluatePosition(piece: Piece): Double {
        // Create a copy of the grid with the piece placed
        val testGrid = grid.copy()
        testGrid.placePiece(piece)
        
        // Calculate evaluation metrics
        val height = testGrid.getAggregateHeight()
        val holes = testGrid.getHoles()
        val bumpiness = testGrid.getBumpiness()
        val wells = testGrid.getWells()
        val lines = countCompleteLines(testGrid)
        val deepHoles = countDeepHoles(testGrid)
        val transitions = countTransitions(testGrid)
        
        // Calculate weighted score
        var score = 0.0
        score += WEIGHTS["height"]!! * height
        score += WEIGHTS["holes"]!! * holes
        score += WEIGHTS["bumpiness"]!! * bumpiness
        score += WEIGHTS["wells"]!! * wells
        score += WEIGHTS["lines"]!! * lines * lines // Square for emphasis
        score += WEIGHTS["deepHoles"]!! * deepHoles
        score += WEIGHTS["transitions"]!! * transitions
        
        // Advanced strategies for higher difficulties
        if (difficultyLevel >= LEVEL_HARD) {
            // Check for T-spin setups
            if (piece.type == PieceType.T && checkTSpinSetup(testGrid, piece)) {
                score += WEIGHTS["tSpinSetup"]!! * 100
            }
            
            // Check for Tetris well
            if (checkTetrisWell(testGrid)) {
                score += WEIGHTS["tetrisWell"]!! * 50
            }
        }
        
        // Combo tracking
        if (comboTracking && lines > 0) {
            score += lines * 10 // Bonus for maintaining combos
        }
        
        return score
    }
    
    private fun evaluateFuture(move: Move, depth: Int): Double {
        if (depth <= 0 || nextPieces.isEmpty()) return 0.0
        
        // Simulate placing the piece
        val testGrid = grid.copy()
        val testPiece = currentPiece?.copy(x = move.x, rotation = move.rotation)
        testPiece?.let {
            testGrid.placePiece(dropPiece(it))
        }
        
        // Evaluate next piece placement
        val nextPiece = Piece(nextPieces.first(), TetrisEngine.GRID_WIDTH / 2, 0, 0)
        val nextMoves = findMovesForPiece(nextPiece)
        
        return nextMoves.maxOfOrNull { it.score } ?: 0.0
    }
    
    private fun isValidPlacement(piece: Piece): Boolean {
        val shape = piece.getCurrentShape()
        
        for (row in shape.indices) {
            for (col in shape[row].indices) {
                if (shape[row][col] != 0) {
                    val gridX = piece.x + col
                    val gridY = piece.y + row
                    
                    if (gridX < 0 || gridX >= TetrisEngine.GRID_WIDTH || 
                        gridY >= TetrisEngine.TOTAL_HEIGHT) {
                        return false
                    }
                    
                    if (gridY >= 0 && grid.getCell(gridX, gridY) != null) {
                        return false
                    }
                }
            }
        }
        
        return true
    }
    
    private fun dropPiece(piece: Piece): Piece {
        var droppedPiece = piece.copy()
        while (isValidPlacement(droppedPiece.copy(y = droppedPiece.y + 1))) {
            droppedPiece.y++
        }
        return droppedPiece
    }
    
    private fun countCompleteLines(testGrid: Grid): Int {
        var lines = 0
        for (y in 0 until TetrisEngine.TOTAL_HEIGHT) {
            var complete = true
            for (x in 0 until TetrisEngine.GRID_WIDTH) {
                if (testGrid.getCell(x, y) == null) {
                    complete = false
                    break
                }
            }
            if (complete) lines++
        }
        return lines
    }
    
    private fun countDeepHoles(testGrid: Grid): Int {
        var deepHoles = 0
        
        for (x in 0 until TetrisEngine.GRID_WIDTH) {
            var blockFound = false
            var holesBelow = 0
            
            for (y in 0 until TetrisEngine.TOTAL_HEIGHT) {
                if (testGrid.getCell(x, y) != null) {
                    blockFound = true
                    if (holesBelow > 2) { // Deep hole is 3+ cells deep
                        deepHoles += holesBelow
                    }
                    holesBelow = 0
                } else if (blockFound) {
                    holesBelow++
                }
            }
        }
        
        return deepHoles
    }
    
    private fun countTransitions(testGrid: Grid): Int {
        var transitions = 0
        
        // Horizontal transitions
        for (y in 0 until TetrisEngine.TOTAL_HEIGHT) {
            for (x in 0 until TetrisEngine.GRID_WIDTH - 1) {
                val current = testGrid.getCell(x, y) != null
                val next = testGrid.getCell(x + 1, y) != null
                if (current != next) transitions++
            }
        }
        
        // Vertical transitions
        for (x in 0 until TetrisEngine.GRID_WIDTH) {
            for (y in 0 until TetrisEngine.TOTAL_HEIGHT - 1) {
                val current = testGrid.getCell(x, y) != null
                val next = testGrid.getCell(x, y + 1) != null
                if (current != next) transitions++
            }
        }
        
        return transitions
    }
    
    private fun checkTSpin(piece: Piece): Boolean {
        if (piece.type != PieceType.T) return false
        
        // Check corners for T-spin detection
        val corners = listOf(
            Pair(piece.x - 1, piece.y - 1),
            Pair(piece.x + 1, piece.y - 1),
            Pair(piece.x - 1, piece.y + 1),
            Pair(piece.x + 1, piece.y + 1)
        )
        
        var filledCorners = 0
        for ((x, y) in corners) {
            if (x < 0 || x >= TetrisEngine.GRID_WIDTH || y >= TetrisEngine.TOTAL_HEIGHT || 
                grid.getCell(x, y) != null) {
                filledCorners++
            }
        }
        
        return filledCorners >= 3
    }
    
    private fun checkTSpinSetup(testGrid: Grid, piece: Piece): Boolean {
        // Check if the current position sets up a future T-spin
        if (difficultyLevel < LEVEL_HARD) return false
        
        // Look for T-spin slots in the grid
        for (y in 1 until TetrisEngine.TOTAL_HEIGHT - 1) {
            for (x in 1 until TetrisEngine.GRID_WIDTH - 1) {
                if (testGrid.getCell(x, y) == null) {
                    // Check if this could be a T-spin slot
                    val corners = listOf(
                        testGrid.getCell(x - 1, y - 1) != null,
                        testGrid.getCell(x + 1, y - 1) != null,
                        testGrid.getCell(x - 1, y + 1) != null,
                        testGrid.getCell(x + 1, y + 1) != null
                    )
                    
                    if (corners.count { it } >= 3) {
                        return true
                    }
                }
            }
        }
        
        return false
    }
    
    private fun checkTetrisWell(testGrid: Grid): Boolean {
        // Check if there's a well suitable for Tetris
        for (x in listOf(0, TetrisEngine.GRID_WIDTH - 1)) {
            var wellDepth = 0
            
            for (y in TetrisEngine.TOTAL_HEIGHT - 1 downTo 0) {
                if (testGrid.getCell(x, y) == null) {
                    wellDepth++
                } else {
                    break
                }
            }
            
            if (wellDepth >= 4) {
                return true
            }
        }
        
        return false
    }
    
    private fun executeMove(move: Move) {
        // This would send commands to the game engine
        // In practice, this would be handled by the Battle mode
    }
    
    fun reset() {
        grid.clear()
        currentPiece = null
        nextPieces = emptyList()
        holdPiece = null
        canHold = true
        linesCleared = 0
        score = 0
        isGameOver = false
    }
    
    fun receiveGarbage(lines: Int) {
        // Handle incoming garbage lines
        // Add gray lines to the bottom of the grid
    }
    
    fun getClearedLines(): Int {
        val cleared = linesCleared
        linesCleared = 0
        return cleared
    }
    
    fun isGameOver(): Boolean {
        return isGameOver
    }
    
    fun getScore(): Int {
        return score
    }
}